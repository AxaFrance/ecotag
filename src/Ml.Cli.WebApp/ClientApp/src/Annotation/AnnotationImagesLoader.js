import React, {useEffect, useState} from "react";
import {fetchGetData, fetchPostJson, utf8_to_b64} from "../FetchHelper";
import {useMutation} from "react-query";
import CroppingLazy from "./Toolkit/BoundingBox/CroppingLazy";
import OcrLazy from "./Toolkit/Ocr/OcrLazy";
import JsonEditorContainer from "./Toolkit/JsonEditor/JsonEditor.container";
import TagOverTextLabelLazy from "./Toolkit/TagOverTextLabel/TagOverTextLabelLazy";
import TagOverTextLazy from "./Toolkit/TagOverText/TagOverTextLazy";
import IrotLazy from "./Toolkit/Rotation/IrotLazy";
import NamedEntityLazy from "./Toolkit/NamedEntity/NamedEntityLazy";
import {toast} from "react-toastify";
import ImageClassifierLazy from "./Toolkit/ImageClassifier/ImageClassifierLazy";


const fetchImages = async data => {
    if (data.status === 200) {
        const hardDriveLocations = await data.json();
        return hardDriveLocations.map(element => {
            return `api/files/${utf8_to_b64(element)}`;
        });
    } else {
        return [];
    }
};

const getImages = (fetchFunction) => async (item) => {
    const params = "fileName=" + item.fileName + "&stringsMatcher="+ item.frontDefaultStringsMatcher + "&directory=" + item.imageDirectory
    const fetchResult = await fetchGetData(fetchFunction)("api/datasets/"+utf8_to_b64(params));
    return fetchImages(fetchResult);
};

const getHttpResultItem = async (item, fetchFunction) => {
    let pathSeparator = "\\";
    if(item.fileDirectory.includes("/")){
        pathSeparator = "/";
    }
   const filePath = `${item.fileDirectory}${pathSeparator}${item.fileName}`
    
    const fetchResult = await fetchGetData(fetchFunction)("api/annotations/"+utf8_to_b64(filePath));
    if (fetchResult.status === 200) {
        const dataObject = await fetchResult.json();
        return {httpResult: dataObject, errorMessage: ""};
    }
    const errorMessage = `Requête invalide: ${fetchResult.statusText}`;
    return {errorMessage, httpResult: {}};
};

const sendConfirmationMessage = (isSuccess) => {
    const message = isSuccess ? "Annotation saved" : "Impossible to save annotation";
    const type = isSuccess ? "success" : "error";
    toast(message, {
        position: "top-right",
        autoClose: 3000,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: false,
        draggable: true,
        progress: undefined,
        type: type
    });
}

const AnnotationImagesLoader = ({item, MonacoEditor, parentState, onSubmit, fetchFunction}) => {

    const [state, setState] = useState({
        fileUrls: [],
        filePrimaryUrl: "",
        httpResultItem: {},
        errorMessage: "",
        isFetched: false
    });

    const mutationDataset = useMutation(newData => fetchPostJson(fetchFunction)("/api/annotations", newData));

    const getUrls = async () => {
        const newUrls = await getImages(fetchFunction)(item);
        const newFileUrl = newUrls != null ? newUrls[0] : "";
        return {fileUrls: newUrls, filePrimaryUrl: newFileUrl};
    };

    const getEditorContent = async () => {
        const {httpResult, errorMessage} = await getHttpResultItem(item, fetchFunction);
        return {errorMessage, httpResultItem: httpResult, isFetched: true};
    };
    
    const getBoundingBoxes = () => {
        let boundingBoxes;
        try{
            const annotationsArray = JSON.parse(item.annotations);
            const lastAnnotation = annotationsArray[annotationsArray.length - 1];
            const labels = lastAnnotation.labels;
            boundingBoxes = labels.boundingBoxes;
        }
        catch (ex){
            boundingBoxes = [];
        }
        return boundingBoxes;
    }

    useEffect(() => {
        let isMounted = true;
        if(parentState.annotationType === "NamedEntityRecognition"){
            return;
        }
        getUrls().then(urls => {
            getEditorContent().then(content => {
                if (isMounted) {
                    setState({
                        fileUrls: urls.fileUrls,
                        filePrimaryUrl: urls.filePrimaryUrl,
                        errorMessage: content.errorMessage,
                        httpResultItem: content.httpResultItem,
                        isFetched: content.isFetched
                    });
                }
            })
        });
        return () => {
            isMounted = false;
        }
    }, []);

    const setAnnotationObject = e => {
        let returnedObject;
        switch (parentState.annotationType) {
            case "JsonEditor":
                returnedObject = {
                    "content": e
                };
                break;
            case "Ocr":
                returnedObject = {
                    "type": e.type,
                    "width": e.width,
                    "height": e.height,
                    "labels": e.labels
                };
                break;
            case "Cropping":
                returnedObject = {
                    "type": e.type,
                    "width": e.width,
                    "height": e.height,
                    "labels": e.labels
                };
                break;
            case "Rotation":
                returnedObject = {
                    "type": e.type,
                    "width": e.width,
                    "height": e.height,
                    "labels": e.labels,
                    "image_anomaly": e.image_anomaly
                }
                break;
            case "TagOverText":
                returnedObject = {
                    "type": e.type,
                    "width": e.width,
                    "height": e.height,
                    "labels": e.labels
                }
                break;
            case "TagOverTextLabel":
                returnedObject = {
                    "type": e.type,
                    "width": e.width,
                    "height": e.height,
                    "labels": e.labels
                };
                break;
            case "NamedEntityRecognition":
                returnedObject = e;
                break;
            case "ImageClassifier":
                returnedObject = {
                    "label": e
                };
                break;
        }
        return returnedObject;
    }

    const onDatasetSubmit = async e => {
        const annotationObject = {
            datasetLocation: parentState.datasetLocation,
            annotationType: parentState.annotationType,
            fileName: item.fileName,
            annotation: setAnnotationObject(e)
        };
        await mutationDataset.mutateAsync(annotationObject, {
            onSuccess: () => sendConfirmationMessage(true),
            onError: () => sendConfirmationMessage(false)
        });
        onSubmit();
    }

    return (
        <>
            {parentState.annotationType === "JsonEditor" &&
            <>
                {!state.isFetched &&
                    <div>Loading editor...</div>
                }
                {state.isFetched &&
                    <JsonEditorContainer
                        expectedOutput={{id: item.id, fileName: item.fileName, value: state.httpResultItem.body}}
                        urls={state.fileUrls}
                        onSubmit={onDatasetSubmit}
                        MonacoEditor={MonacoEditor}
                    />
                }
                {state.errorMessage &&
                    <div className="error-message">{state.errorMessage}</div>
                }
            </>
            }
            {parentState.annotationType === "Ocr" &&
            <OcrLazy
                labels={parentState.configuration}
                expectedLabels={[]}
                url={state.filePrimaryUrl}
                onSubmit={onDatasetSubmit}
            />
            }
            {parentState.annotationType === "Cropping" &&
            <CroppingLazy
                labels={parentState.configuration}
                url={state.filePrimaryUrl}
                onSubmit={onDatasetSubmit}
            />
            }
            {parentState.annotationType === "Rotation" &&
                <IrotLazy
                    expectedLabels={[]}
                    url={state.filePrimaryUrl}
                    onSubmit={onDatasetSubmit}
                />
            }
            {parentState.annotationType === "TagOverText" &&
                <TagOverTextLazy
                    expectedOutput={getBoundingBoxes()}
                    url={state.filePrimaryUrl}
                    onSubmit={onDatasetSubmit}
                />
            }
            {parentState.annotationType === "TagOverTextLabel" &&
                <TagOverTextLabelLazy
                    expectedOutput={getBoundingBoxes()}
                    url={state.filePrimaryUrl}
                    onSubmit={onDatasetSubmit}
                    labels={parentState.configuration}
                />
            }
            {parentState.annotationType === "NamedEntityRecognition" &&
                <NamedEntityLazy
                    text={"Hello, my name is Lilian Delouvy, and this is some sample text.  The NER is currently in a 'Work In Progress' state. And it will be awesome. bla bla . bla bla . bla bla . bla bla . bla bla "}
                    labels={parentState.configuration}
                    annotationAction={onDatasetSubmit}
                    placeholder="Submit Annotation"
                />
            }
            {parentState.annotationType === "ImageClassifier" &&
                <ImageClassifierLazy
                    url={state.filePrimaryUrl}
                    labels={parentState.configuration}
                    onSubmit={onDatasetSubmit}
                />
            }
        </>
    );
};

export default AnnotationImagesLoader;
