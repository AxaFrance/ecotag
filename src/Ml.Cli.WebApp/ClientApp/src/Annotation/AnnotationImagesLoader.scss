.message{
  width: 100%;
  text-align: center;
  margin-bottom: 1em;
  &--success{
    color: green;
  }
  &--error{
    color: red;
  }
}