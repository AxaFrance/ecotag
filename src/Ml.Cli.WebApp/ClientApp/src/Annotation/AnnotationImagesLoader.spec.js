import {render, waitFor} from "@testing-library/react";
import React from "react";
import AnnotationImagesLoader from "./AnnotationImagesLoader";
import {Textarea} from "@axa-fr/react-toolkit-form-input-textarea";
import {QueryClient, QueryClientProvider} from "react-query";

const mockedMonacoEditor = ({id, value, onChange}) => {
    return (
        <Textarea
            id={id}
            name="Textarea"
            value={value}
            onChange={onChange}
        />
    );
};

const mockedFetchFunction = () => {
    return {
        ok: true,
        status: 200,
        json: async () => ["C:\\\\imageLocation"]
    };
};

describe("Check dataset handling", () => {
    test("Should render AnnotationImagesLoader and get image", async () => {
        const queryClient = new QueryClient();
        const configuration = JSON.parse("[{\"name\": \"Recto\", \"id\": 0}, {\"name\": \"Verso\", \"id\": 1}]");
        const {container, asFragment, getByAltText} = render(
            <QueryClientProvider client={queryClient}>
                <AnnotationImagesLoader
                    item={{id: "annotation_image_loader_editor_id", fileName: "fileName.json", imageDirectory: "imageDirectoryValue"}}
                    MonacoEditor={mockedMonacoEditor}
                    parentState={{configuration, annotationType: "JsonEditor"}}
                    fetchFunction={mockedFetchFunction}
                />
            </QueryClientProvider>
        );
        await waitFor(() => expect(getByAltText(/file_image/i)).not.toBeNull());
        const image = container.querySelector("img");
        expect(image.getAttribute("src")).toEqual("api/files/value=C%3A%5C%5CimageLocation");
        expect(asFragment()).toMatchSnapshot();
    });
});
