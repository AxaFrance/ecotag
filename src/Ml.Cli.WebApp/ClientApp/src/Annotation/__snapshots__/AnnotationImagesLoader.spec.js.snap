// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check dataset handling Should render AnnotationImagesLoader and get image 1`] = `
<DocumentFragment>
  <div
    class="images-editor-container"
  >
    <div
      class="images-editor-container__images_column"
    >
      <div
        class="images-list"
      >
        <div
          class="images-list__item"
        >
          <div
            class="images-list__image-container"
          >
            <img
              alt="file_image"
              class="images-list__file_image"
              src="api/files/value=C%3A%5C%5CimageLocation"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="editor-container"
    >
      <textarea
        class="af-form__input-textarea"
        id="annotation_image_loader_editor_id"
        name="Textarea"
        placeholder=""
      />
      <div
        class="editor-container__errors-container"
      >
        <div
          class="editor-container__error-message"
        >
          <span>
            Impossible de parser l'item fileName.json. Veuillez vous assurer que le contenu de ce fichier est au format JSON.
          </span>
        </div>
      </div>
      <div
        class="editor-container__button"
      >
        <button
          class="btn af-btn"
        >
          Sauvegarder
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
